### ISSUE TITLE
[SDK 50]  Function calls prefixed with `void` together with optional chain (`?.`) are skipped

### ISSUE THREAD

**Issue by Gamote** [expo/expo#26276]
### Minimal reproducible example

https://github.com/Gamote/expo-babel-void-optional-chain-issue

## Description

It seems that `void someFunction?.()` skips the function call in production builds.

I have checked the transpiled code using the debugger and `void someFunction?.()` is transpiled to `undefined` in production builds.

If we do `someFunction?.()` or `void someFunction()` it will work just fine.

### Additional information

- I have tested this on iOS and Android
- I have tested this on SDK 49 and it works as expected
- When starting with `expo start` it works as expected

## Reproduce the issue

1. Clone the repo
2. (optional) Run `nvm use`
3. Run `yarn`
4. Start the app `yarn ios --no-dev` or `yarn android --no-dev`

## Expected behaviour

Three alerts should be shown, one after the other.

## Actual behaviour

Only two alerts are shown.

https://github.com/expo/expo/assets/38785171/7e0a780b-5832-443c-bb72-2edbdb802b6e

### Environment

    System:
      OS: macOS 14.2.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
      Yarn: 1.22.21 - ~/.nvm/versions/node/v20.10.0/bin/yarn
      npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
      Watchman: 2023.12.04.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
    IDEs:
      Android Studio: 2023.1 AI-231.9392.1.2311.11255304
      Xcode: 15.0.1/15A507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^50.0.0-preview.7 => 50.0.0-preview.7 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.1 => 0.73.1 
    npmGlobalPackages:
      eas-cli: 5.9.3
    Expo Workflow: managed

**Comment by Gamote** [comment]
Ok took me a while (first time I interact with the `expo` code base 😄) but managed to narrow it down to the `@expo/metro-config` packages.

It seems that `transformJSWithBabel` ([src/transform-worker/metro-transform-worker.ts#L434](https://github.com/expo/expo/blob/ae7bb2f350199e622fb92e6bbda0a02cb980eda9/packages/%40expo/metro-config/src/transform-worker/metro-transform-worker.ts#L434)) is calling `transformJS`  ([src/transform-worker/metro-transform-worker.ts#L167](https://github.com/expo/expo/blob/ae7bb2f350199e622fb92e6bbda0a02cb980eda9/packages/%40expo/metro-config/src/transform-worker/metro-transform-worker.ts#L167)) which then returns the JS code in which `void someFunction?.()` is converted into `undefined`.

Before that everything is alright and the original data about the correct call is preserved.

It seems that this issue was introduced together with the update to `@expo/cli@^16.x.x`.

I am still looking into and, hopefully I can find the exact cause and create a PR.

**Comment by Gamote** [comment]
Narrowed it down a bit more.

It seems that the issue is when the `ast` is transformed using the `constantFoldingPlugin` ([line](https://github.com/expo/expo/blob/ae7bb2f350199e622fb92e6bbda0a02cb980eda9/packages/%40expo/metro-config/src/transform-worker/metro-transform-worker.ts#L258)) on **non-dev** environments:
```
...
plugins: [[_metroTransformPlugins().default.constantFoldingPlugin, babelPluginOpts]],
...
```

**Comment by Gamote** [comment]
Ok so I am inclined to think that this is an issue for `metro-transform-plugins` package, as I found a way to fix it on their codebase. See more: https://github.com/facebook/metro/issues/1167.

I am still not sure if there is anything extra that needs to be done on Expo's side. The reason I am saying this is because I saw that the code on `constant-folding-plugin` didn't change in years and I can't reproduce the issue in SDK 49, so either:
- `metro-transform-plugins` wasn't used before SDK 50
- or there is something else that I am missing (being quite new to the Expo codebase)

Please let me know your thoughts on this, and I will be more than happy to help towards a resolution.

**Comment by brentvatne** [comment]
@Gamote - @EvanBacon is the owner of this area of the codebase and is still out on winter holiday, he'll be back next week and can look into this. One possibility here is that we could stop using the `constant-folding-plugin`

**Comment by Gamote** [comment]
> @Gamote - @EvanBacon is the owner of this area of the codebase and is still out on winter holiday, he'll be back next week and can look into this. One possibility here is that we could stop using the `constant-folding-plugin`
> 
> edit: in the meantime, I pinged the Metro team about this

Thank you @brentvatne. Will wait for @EvanBacon.

About the "stop using the plugin" part, maybe it will not be necessary if the direction of this PR is the correct one https://github.com/facebook/metro/pull/1178. Not sure, will wait for the Metro team's feedback.

**Comment by Gamote** [comment]
The PR was merged in, I will wait for a release, test, and come back with feedback on the matter.

**Comment by Gamote** [comment]
Hi @EvanBacon, the fix for this issue was released on Metro team's side ([metro@v0.80.5](https://github.com/facebook/metro/releases/tag/v0.80.5)).

Whenever you have time, can you please let me know if a PR with the version change is what I should do next?

**Comment by Gamote** [comment]
I think this issue can be closed. I did a `yarn upgrade` and the `metro-transform-plugins` got updated to `0.80.5` and everything works now.
